reduce the number of translatable strings in the unit type editor

babaissarkar opened the pull request:
Resolves #8677.

babaissarkar commented:
Note that this leads to some inconsistent spacing and colons (:).

CelticMinstrel commented:
Just checking, but you're aware that a textdomain holds until the next textdomain, right? Since there are some strings now that don't have a textdomain above them but are now loaded from a different textdomain than before.

(I was actually wondering if it would make sense to add a redundant textdomain declaration above them to make it clearer…? Or is that silly?)

babaissarkar commented:
Yes, I know that the textdomain holds until the next one is declared. I'll
try to add extra declarations to make the intent clear.

On Thu 11 Apr, 2024, 10:12 AM Celtic Minstrel, ***@***.***>
wrote:

> Just checking, but you're aware that a textdomain holds until the next
> textdomain, right? Since there are some strings now that don't have a
> textdomain above them but are now loaded from a different textdomain than
> before.
>
> (I was actually wondering if it would make sense to add a redundant
> textdomain declaration above them to make it clearer…? Or is that silly?)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wesnoth/wesnoth/pull/8731#issuecomment-2048915737>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAT3ADDIHOIFV56YP3ZAQTY4YIC3AVCNFSM6AAAAABGBS3G7CVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBYHEYTKNZTG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

stevecotton commented:
We've got a long time before 1.20 releases. Unless this is specifically aiming for a backport to 1.18, I think it would be worth waiting to see if RTL support causes changes to the strings that end "colon space".

babaissarkar commented:
> We've got a long time before 1.20 releases. Unless this is specifically aiming for a backport to 1.18, I think it would be worth waiting to see if RTL support causes changes to the strings that end "colon space".

So there's work ongoing on RTL support?

Some more explanation would be good. What would you suggest? postpone this until later?
I don't have any problems with backporting this to 1.18 if this helps the translators.

stevecotton commented:
I mean, the unit editor isn't in 1.18, so these strings would only affect 1.18 if the unit editor itself was going to be backported - I remember discussing that, not sure if it's still happening or not.

While there's a lot of RTL UI improvements that could still be done, the Arabic translation is now over 80% complete, and [complete for every campaign except EI and TSG](https://gettext.wesnoth.org/?view=langs&version=branch&lang=ar). That raised the priority of #5600, and the help is now in a usable state, although the right-hand pane of the help shows the problems with having two labels next to each other, "heading: " "value", when the positioning of the labels expects an LTR language. I think the solution is going to be "heading: $value", so that the text direction works.

babaissarkar commented:
> "heading: $value"
That's a good idea. Although Bengali is LTR, it still suffers from the same problem time to time, where having the values as variables would help.

I should probably keep RTL in mind for #8655, though the main issue lies in the help topic generators probably.

Vultraz commented:
Why does this add spaces after colons in some cases?

stevecotton commented:
@babaissarkar: if I understand correctly, for Bengali grammar you need something like "Becomes $value when promoted"; so you'll also want the coloration markup to be part of the translatable string, so that the unit name can have white text both before and after?

@Vultraz: because it's trying to reuse existing strings, and some of the existing ones have a space after the colon.

babaissarkar commented:
@stevecotton 
> so you'll also want the coloration markup to be part of the translatable string, so that the unit name can have white text both before and after?
Yes, that's definitely needed if parts of the string are formatted differently. Can only be ignored if the entire string is formatted in the same way. So for `<i>I am an unit</i>` I don't need the markup, but for `I am an <i>unit</i>`, I need it, because 'unit' might be placed somewhere else in the translated version.

babaissarkar commented:
@CelticMinstrel That's also fine. If the translation can work without having to bother about formatting, that's even better, but it might require more reorganizing in the codebase.

babaissarkar commented:
2f4d6bafe2098e56996151e8034156a6691cdb6e also contains a couple of small bugfixes.
@stevecotton is it fine to mark the alignment strings as untranslatable? they are basically values that are inserted directly into the WML. Look at this :
![Screenshot from 2024-04-16 20-14-41](https://github.com/wesnoth/wesnoth/assets/8469888/653782ae-88dc-4e90-83c6-9db48fe27878)

Another weird issue :
![Screenshot from 2024-04-16 20-15-21](https://github.com/wesnoth/wesnoth/assets/8469888/1b6e2e49-126d-4fe3-8813-ced62ebfa671)

Translated text gets inserted into the WML when using a different language.

soliton- commented:
That's a bug. WML code should not be translated.

babaissarkar commented:
Bugfixes moved to #8846 including https://github.com/wesnoth/wesnoth/pull/8731#issuecomment-2059282680.

Pentarctagon commented:
Is there anything remaining for this PR aside from resolving the conflicts?

babaissarkar commented:
> Is there anything remaining for this PR aside from resolving the conflicts?

Probably not.
@stevecotton Are the changes in this PR still relevant? I think most of the translation teams have already translated the new strings in Unit Type Editor, in which case this becomes unnecessary.

stevecotton commented:
> Are the changes in this PR still relevant? I think most of the translation teams have already translated the new strings in Unit Type Editor, in which case this becomes unnecessary.

I'm confused - AFAICS, the Unit Type Editor is only in 1.19. The translation teams generally only work on 1.18.

babaissarkar commented:
@stevecotton can this be merged, or you'd suggest waiting some more?

stevecotton commented on src/gui/dialogs/editor/edit_unit.cpp:
```diff
@@ -69,7 +66,7 @@ editor_edit_unit::editor_edit_unit(const game_config_view& game_config, const st
 	read(specials, *(preprocess_file(game_config::path+"/data/core/macros/abilities.cfg", &abilities_map_)));
 	for (const auto& x : abilities_map_) {
-		// Don't add any macros that have INTERNAL
+		/** Don't add any macros that have INTERNAL */
```
This isn't the place for a Doxygen comment, it should either stay as `//` or be `/* .... */`.

babaissarkar replied:
Updated.

stevecotton commented:
Thinking about 1.20, for non-editable values, I think this should use `heading: $value`, with coloration included in `$value`, both ideas as said above.

But, when the translatable string is a label next to an editbox, I guess it needs a version that doesn't have the `$value`, and shouldn't have a trailing space (and just assume that it's in the editor, so not worry about bidirectional text for this). My feeling is to ask @irydacea whether to put such strings in `wesnoth` or `wesnoth-lib`, but to get a complete set in one domain.

soliton- commented on src/gui/dialogs/editor/edit_unit.hpp:
```diff
@@ -58,6 +58,8 @@ class editor_edit_unit : public modal_dialog
 	std::vector<config> align_list_, race_list_, movetype_list_, defense_list_, resistances_list_, usage_type_list_;
 	std::vector<config> specials_list_, abilities_list_;
+	// hardcoded list of untranslatated alignment strings
+	std::vector<std::string> align_list_untrans_ = {"lawful", "chaotic", "neutral", "liminal"};
```
Why can't you use `unit_alignments::values`?

babaissarkar replied:
Didn't know about that, updated.

Wedge009 commented:
Where do we stand here? Looks like it's pretty close to merging - if not already there - but I don't know if there are still outstanding concerns.

babaissarkar commented:
> Where do we stand here? Looks like it's pretty close to merging - if not already there - but I don't know if there are still outstanding concerns.

Waiting for @stevecotton's approval.

Vultraz commented:
Looking at this again, it doesn’t seem great that some labels use colons and others do not. Personally, I’d say don’t merge.

babaissarkar commented:
> Looking at this again, it doesn’t seem great that some labels use colons and others do not. Personally, I’d say don’t merge.

I think I agree that inconsistent colons do lead to bad UI design, and we probably need a couple more strings with colons. What about the non-colon changes, or if I make the colons consistent? (though we might need some more string)

stevecotton commented on src/gui/dialogs/editor/edit_unit.cpp:
```diff
@@ -90,11 +89,11 @@ void editor_edit_unit::pre_show(window& win) {
-	align_list_.emplace_back("label", _("chaotic"));
-	align_list_.emplace_back("label", _("neutral"));
-	align_list_.emplace_back("label", _("liminal"));
+	for (size_t i = 0; i < unit_alignments::values.size(); i++) {
```
This could be a range-based for loop.

babaissarkar replied:
Updated.

stevecotton commented:
Clarifying a previous comment of mine, we shouldn't have trailing spaces in `"Property:"` strings, even if that cleanup means that we have near-duplicates that differ only by the trailing space for a while.

I'm still not sure whether they should be in `wesnoth` or `wesnoth-lib`, but we should have the complete set of strings for showing the properties of a unit, all in one text domain or the other.

stevecotton commented on data/gui/themes/default/window/editor_edit_unit.cfg:
```diff
@@ -251,8 +252,10 @@
+												#textdomain wesnoth-lib
+												{_GUI_CHECK_BUTTON gender_male ( _ "Male")}
+												{_GUI_CHECK_BUTTON gender_female ( _ "Female")}
+												#textdomain wesnoth-help
```
This line seems to be a leftover

stevecotton commented on data/gui/themes/default/window/editor_edit_unit.cfg:
```diff
@@ -218,7 +218,8 @@
 											definition = "default"
-											label = _ "Level:"
+											#textdomain wesnoth
+											label = _ "Level: "
```
I'm against having trailing spaces in strings like this. Even if that currently means that they can't reuse a string, the eventual aim of the cleanup should be to remove the trailing space from the existing translatable strings.

babaissarkar replied:
Sure

babaissarkar replied:
corresponding commit dropped.

babaissarkar commented:
@stevecotton I can move this entirely into the `wesnoth` textdomain and remove the trailing spaces.

babaissarkar commented:
> At least for the moment, my feeling is to take the commit that uses the existing unit_alignment data, but drop the commit with the string changes.

I've dropped that commit (and replaced another occurrence of `vgettext_impl`)

stevecotton commented on src/gui/dialogs/editor/edit_unit.cpp:
```diff
@@ -109,11 +108,11 @@ void editor_edit_unit::pre_show(window& win) {
 	button& load = find_widget<button>(&win, "load_unit_type", false);
 	std::stringstream tooltip;
-	tooltip << vgettext_impl("wesnoth", "Hotkey(s): ",  {{}});
+	tooltip << t_string("Hotkey(s): ", "wesnoth");
```
This (and the existing implementation) only get into the .po files because the same string is in theme.cpp, but it's not a regression so let's not worry about it for now.
